autorag/nodes/passagefilter/percentile_cutoff.py

```diff
@@ -0,0 +1,95 @@
+import itertools
+from typing import List, Tuple, Optional
+
+import numpy as np
+import torch.cuda
+
+from autorag import embedding_models
+from autorag.evaluate.metric.util import calculate_cosine_similarity
+from autorag.nodes.passagefilter.base import passage_filter_node
+from autorag.utils.util import reconstruct_list
+
+
+@passage_filter_node
+def similarity_percentile_cutoff(queries: List[str], contents_list: List[List[str]],
+                                 scores_list: List[List[float]], ids_list: List[List[str]],
+                                 percentile: float, embedding_model: Optional[str] = None,
+                                 batch: int = 128,
+                                 ) -> Tuple[List[List[str]], List[List[str]], List[List[float]]]:
+    """
+    Re-calculate each content's similarity with the query and filter out the contents that are below the content's
+    length times percentile. If This is a filter and does not override scores. The output of scores is not coming from
+    query-content similarity.
+    If the value of content's length times percentile is less than 1, keep the only one highest similarity content.
+
+    :param queries: The list of queries to use for filtering
+    :param contents_list: The list of lists of contents to filter
+    :param scores_list: The list of lists of scores retrieved
+    :param ids_list: The list of lists of ids retrieved
+    :param percentile: The percentile to cut off
+    :param embedding_model: The embedding model to use for calculating similarity
+        Default is OpenAIEmbedding.
+    :param batch: The number of queries to be processed in a batch
+        Default is 128.
+    :return: Tuple of lists containing the filtered contents, ids, and scores
+    """
+    if embedding_model is None:
+        embedding_model = embedding_models['openai']
+    else:
+        embedding_model = embedding_models[embedding_model]
+
+    # Embedding using batch
+    embedding_model.embed_batch_size = batch
+    query_embeddings = embedding_model.get_text_embedding_batch(queries)
+
+    content_lengths = list(map(len, contents_list))
+    content_embeddings_flatten = embedding_model.get_text_embedding_batch(list(
+        itertools.chain.from_iterable(contents_list)))
+    content_embeddings = reconstruct_list(content_embeddings_flatten, content_lengths)
+
+    results = list(map(lambda x: similarity_percentile_cutoff_pure(x[0], x[1], x[2], x[3], x[4], percentile),
+                       zip(query_embeddings, content_embeddings, contents_list, ids_list, scores_list)))
+
+    remain_content_list = list(map(lambda x: x[0], results))
+    remain_ids_list = list(map(lambda x: x[1], results))
+    remain_scores_list = list(map(lambda x: x[2], results))
+
+    del embedding_model
+    if torch.cuda.is_available():
+        torch.cuda.empty_cache()
+
+    return remain_content_list, remain_ids_list, remain_scores_list
```

Comment:
Looks like duplicate code from similarity_threshold_cutoff.
Why don't you make a external function for reducing repeat code?